runtime/wasm/src/module/mod.rs

```diff
@@ -269,126 +125,370 @@ impl WasmiModule {
+                }
+            });
+        }
+
+        *shared_instance.borrow_mut() = Some(WasmInstance {
+            instance,
+            memory_allocate: Box::new(memory_allocate),
+            memory,
+            ctx,
+            valid_module,
+            host_metrics,
+            timeout,
+            should_interrupt,
+
+            // `arena_start_ptr` will be set on the first call to `raw_new`.
+            arena_free_size: 0,
+            arena_start_ptr: 0,
+        });
 
-        result.map(|_| self.ctx.state).map_err(|err| {
-            format_err!(
-                "Failed to handle Ethereum block with handler \"{}\": {}",
-                handler_name,
-                format_wasmi_error(err)
-            )
+        Ok(WasmInstanceHandle {
+            instance: shared_instance.clone(),
         })
     }
 }
 
-impl AscHeap for WasmiModule {
-    fn raw_new(&mut self, bytes: &[u8]) -> Result<u32, Error> {
+impl AscHeap for WasmInstance {
+    fn raw_new(&mut self, bytes: &[u8]) -> u32 {
         // We request large chunks from the AssemblyScript allocator to use as arenas that we
         // manage directly.
 
-        static MIN_ARENA_SIZE: u32 = 10_000;
+        static MIN_ARENA_SIZE: i32 = 10_000;
 
-        let size = u32::try_from(bytes.len()).unwrap();
+        let size = i32::try_from(bytes.len()).unwrap();
         if size > self.arena_free_size {
             // Allocate a new arena. Any free space left in the previous arena is left unused. This
             // causes at most half of memory to be wasted, which is acceptable.
             let arena_size = size.max(MIN_ARENA_SIZE);
-            let allocated_ptr = self
-                .module
-                .clone()
-                .invoke_export("memory.allocate", &[RuntimeValue::from(arena_size)], self)
-                .expect("Failed to invoke memory allocation function")
-                .expect("Function did not return a value")
-                .try_into::<u32>()
-                .expect("Function did not return u32");
-            self.arena_start_ptr = allocated_ptr;
+            self.arena_start_ptr = (self.memory_allocate)(arena_size).unwrap();
             self.arena_free_size = arena_size;
         };
 
-        let ptr = self.arena_start_ptr;
-        self.memory.set(ptr, bytes)?;
+        let ptr = self.arena_start_ptr as usize;
+
+        // Safe because we are accessing and immediately dropping the reference to the data.
+        unsafe { self.memory.data_unchecked_mut()[ptr..(ptr + bytes.len())].copy_from_slice(bytes) }
         self.arena_start_ptr += size;
         self.arena_free_size -= size;
 
-        Ok(ptr)
+        ptr as u32
     }
 
-    fn get(&self, offset: u32, size: u32) -> Result<Vec<u8>, Error> {
-        self.memory.get(offset, size as usize)
-    }
-}
-
-impl<E> HostError for HostExportError<E> where E: fmt::Debug + fmt::Display + Send + Sync + 'static {}
+    fn get(&self, offset: u32, size: u32) -> Vec<u8> {
```

Comment:
Are we sure that this is deterministic? I think it probably is, but I want to be sure. Imagine that a subgraph reads from a specific memory address that was not given to it by a previous call to `raw_new` . Does WebAssembly guarantee that all of the memory layout be deterministic? Will something like upgrading the wasmtime dependency possibly affect that memory layout? Is some of the memory returned by `instance.get_memory("memory")` controlled by wasmtime? Where did that name come from and why didn't we use `Memory::new()` to limit this to just the arena?

Reply:
Can you access the stack with this memory? Can the WebAssembly compiler introduce optimizations that would eliminate dead code and thereby parts of the stack that might be read by some other function.

Reply:
I went digging to find good answers to your questions.

The [store](https://www.w3.org/TR/wasm-core-1/#store) represents the global state of a Wasm module, this includes the memory. In the future there may be multiple memories, but currently there is [only one memory per module](https://github.com/WebAssembly/design/blob/master/Modules.md#linear-memory-section). And at least AS calls it `"memory"`. There exists the option of having AS import a memory that we initialize ourselves instead of letting it initialize its own memory, but that's not what we've done so far and I'm sure what difference that makes. There is no uninitialized memory in Wasm, memory is [zeroed when initialized or grown](https://www.w3.org/TR/wasm-core-1/#grow-mem).

The Wasm stack is separate from the store and its memory, so it cannot be accessed by the host.

Wasm is deterministic, and what we're doing from the host here is no different from what a Wasm [load or store instruction](https://www.w3.org/TR/wasm-core-1/#syntax-instr-memory). I see no indication that accessing memory from the host should be any different from accessing it from Wasm.

And to your point on optimizations, Cranelift is a general-purpose optimizing compiler, and in principle that is scary for determinism. As a minor point the optimization level is set to the default of `None`, so dead code elimination isn't performed, but that could change. The important thing is that even a general purpose optimizing compiler should not violate the spec when compiling Wasm. From what I could see, Cranelift considers [any store to be a side effect](https://github.com/bytecodealliance/wasmtime/blob/master/cranelift/codegen/src/inst_predicates.rs#L19) and therefore not a candidate for [dead code elimination](https://github.com/bytecodealliance/wasmtime/blob/master/cranelift/codegen/src/dce.rs#L23). I did find one mention of dead store removal, but I think that's just a mistake in the comment https://github.com/bytecodealliance/wasmtime/pull/1828.

All that said, there is an alternative backend for wasmtime called Lightbeam, which is overall a more conservative backend and is what Substrate uses. I didn't try it because it doesn't support nan canonicalization.

Reply:
I am satisfied that this is fully deterministic. Consider putting some of your learnings in the comment.

Reply:
I've added some comments around memory and cranelift, and also explicitly set `OptLevel::None`.